src/utils/mod.rs

```diff
@@ -246,30 +247,54 @@ async fn download_file_(
     // Download the file
 
     // Keep the curl env var around for a bit
-    let use_curl_backend = process
-        .var_os("RUSTUP_USE_CURL")
-        .is_some_and(|it| it != "0");
-    let use_rustls = process
-        .var_os("RUSTUP_USE_RUSTLS")
-        .is_none_or(|it| it != "0");
-    let (backend, notification) = if use_curl_backend {
-        (Backend::Curl, Notification::UsingCurl)
-    } else {
-        let tls_backend = if use_rustls {
-            TlsBackend::Rustls
-        } else {
-            #[cfg(feature = "reqwest-native-tls")]
-            {
-                TlsBackend::NativeTls
+    let use_curl_backend = process.var_os("RUSTUP_USE_CURL").map(|it| it != "0");
+    let use_rustls = process.var_os("RUSTUP_USE_RUSTLS").map(|it| it != "0");
+
+    let backend = match (use_curl_backend, use_rustls) {
+        // If environment specifies a backend that's unavailable, error out
+        #[cfg(not(feature = "reqwest-rustls-tls"))]
+        (_, Some(true)) => {
+            return Err(anyhow!(
+                "RUSTUP_USE_RUSTLS is set, but reqwest-rustls-tls is not enabled"
+            ));
+        }
+        #[cfg(not(feature = "reqwest-native-tls"))]
+        (_, Some(false)) => {
+            return Err(anyhow!(
+                "RUSTUP_USE_RUSTLS is set to false, but reqwest-native-tls is not enabled"
```

Comment:
I think the wording can be improved here, since "reqwest-native-tls is not enabled" doesn't necessarily lead me to compile-time configuration.

What about "this rustup distribution was not built with the `reqwest-native-tls` optional feature"?

Reply:
Expanded these to be more explicit.